### PR TITLE
Try to loosen cfunction optimization type check

### PR DESCRIPTION
**Pull request by yuyichao**
Not sure if this is precise now but the old one seems to be too strict since it doesn't catch passing by ref or non-leaf argument types.

### CODE REVIEW

**Comment by tkelman** [comment]
can whatever was wrong with the previous version be tested for?

**Comment by yuyichao** [comment]
Nothing was wrong. It's only a performance issue.

File: src/ccall.cpp [hunk 1]
```diff
@@ -1785,18 +1785,26 @@ static jl_cgval_t emit_ccall(jl_value_t **args, size_t nargs, jl_codectx_t *ctx)
...
-                    fargt = (jl_value_t*)jl_apply_tuple_type_v((jl_value_t**)jl_data_ptr(fargt), jl_nfields(fargt));
-                }
-            }
-            else {
+            if (!fargt) {
                 fargt = expr_type(args[8], ctx);
-                if (jl_is_type_type((jl_value_t*)fargt))
+                if (jl_is_type_type((jl_value_t*)fargt)) {
                     fargt = jl_tparam0(fargt);
+                    if (jl_has_free_typevars(fargt)) {
```
Comment (vtjnash) [inline src/ccall.cpp]: `|| !jl_is_tuple_type(fargt)`

Comment (yuyichao) [inline src/ccall.cpp]: Done.

**Comment by yuyichao** [comment]
Any other comments?

**Comment by tkelman** [comment]
worth running nanosoldier?

**Comment by yuyichao** [comment]
Kind of doubt it has anything that uses cfunction but sure....


@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21467#issuecomment-304097102) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/93b8c91fd67ced091870f4204ed4248db36e9f8e/2f9e3db_vs_4158b5e/report.md). cc @jrevels

**Pull request merged**